extension-style/src/main/java/com/mapbox/maps/extension/style/layers/LayerExt.kt

```diff
@@ -18,23 +19,30 @@ import com.mapbox.maps.logE
  * @return StyleLayerPlugin
  */
 fun StyleManagerInterface.getLayer(layerId: String): Layer? {
-  return this.getStyleLayerProperty(layerId, "type").silentUnwrap<String>()?.let { type ->
-    when (type) {
-      "background" -> BackgroundLayer(layerId).also { it.delegate = this }
-      "location-indicator" -> LocationIndicatorLayer(layerId).also { it.delegate = this }
-      "sky" -> SkyLayer(layerId).also { it.delegate = this }
-      "circle" -> CircleLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      "fill-extrusion" -> FillExtrusionLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      "fill" -> FillLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      "heatmap" -> HeatmapLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      "hillshade" -> HillshadeLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      "line" -> LineLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      "raster" -> RasterLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      "symbol" -> SymbolLayer(layerId, this.getStyleLayerProperty(layerId, "source").unwrap()).also { it.delegate = this }
-      else -> {
-        logE(TAG, "Layer type: $type unknown.")
-        null
-      }
+  fun source(): String = getStyleLayerProperty(layerId, "source").unwrap()
+  val type = getStyleLayerProperty(layerId, "type").silentUnwrap<String>()
+  return when (type) {
+    "background" -> BackgroundLayer(layerId)
+    "location-indicator" -> LocationIndicatorLayer(layerId)
+    "sky" -> SkyLayer(layerId)
+    "circle" -> CircleLayer(layerId, source())
+    "fill-extrusion" -> FillExtrusionLayer(layerId, source())
+    "fill" -> FillLayer(layerId, source())
+    "heatmap" -> HeatmapLayer(layerId, source())
+    "hillshade" -> HillshadeLayer(layerId, source())
+    "line" -> LineLayer(layerId, source())
+    "raster" -> RasterLayer(layerId, source())
+    "symbol" -> SymbolLayer(layerId, source())
+    else -> {
+      logE(TAG, "Layer type: $type unknown.")
+      null
+    }
+  }?.also { result ->
+    result.delegate = this
+    try {
+      result.cachedLayerPropertiesValue = getStyleLayerProperties(layerId).value
+    } catch (e: ClassCastException) {
```

Comment:
Probably not worth it but, we could avoid calling three times the native side by calling `getStyleLayerProperties(...)` in the beginning and then get the required properties directly.

Something like this:
```suggestion
  val layerProperties = getStyleLayerProperties(layerId).value
  val styleLayerProperties: Map<String, Any> = layerProperties!!.unwrapToTyped()
  val source: String? = styleLayerProperties["type"] as String?
  val type = styleLayerProperties["type"]!! as String
  return when (type) {
    "background" -> BackgroundLayer(layerId)
    "location-indicator" -> LocationIndicatorLayer(layerId)
    "sky" -> SkyLayer(layerId)
    "circle" -> CircleLayer(layerId, source!!)
    "fill-extrusion" -> FillExtrusionLayer(layerId, source!!)
    "fill" -> FillLayer(layerId, source!!)
    "heatmap" -> HeatmapLayer(layerId, source!!)
    "hillshade" -> HillshadeLayer(layerId, source!!)
    "line" -> LineLayer(layerId, source!!)
    "raster" -> RasterLayer(layerId, source!!)
    "symbol" -> SymbolLayer(layerId, source!!)
    else -> {
      logE(TAG, "Layer type: $type unknown.")
      null
    }
  }?.also { result ->
    result.delegate = this
    try {
      result.cachedLayerPropertiesValue = layerProperties
    } catch (e: ClassCastException) {
```

Reply:
It's a fair point but think we don't need to go that far at the moment, also tests would need further refactoring since there are often not real values provided and !! causes the test to fail.